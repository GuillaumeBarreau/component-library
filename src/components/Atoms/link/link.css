.link--notice-default {
  color: var(--color-notice-default--text);
}

.link--notice-error {
  color: var(--color-notice-error--default);
}

.link--notice-success {
  color: var(--color-notice-success--default);
}

.link--notice-warning {
  color: var(--color-notice-warning--default);
}

.link--notice-info {
  color: var(--color-notice-info--default);
}

.link--notice-disabled {
  color: var(--color-base-02--dark);
  text-decoration: none;
  cursor: not-allowed;
}

.link--notice-default:focus {
  color: var(--color-notice-default--light);
}

.link--notice-error:focus {
  color: var(--color-notice-error--light);
}

.link--notice-success:focus {
  color: var(--color-notice-success--light);
}

.link--notice-warning:focus {
  color: var(--color-notice-warning--light);
}

.link--notice-info:focus {
  color: var(--color-notice-info--light);
}

.link--notice-default:hover {
  color: var(--color-text--dark);
}

.link--notice-error:hover {
  color: var(--color-notice-error--dark);
}

.link--notice-success:hover {
  color: var(--color-notice-success--dark);
}

.link--notice-warning:hover {
  color: var(--color-notice-warning--dark);
}

.link--notice-info:hover {
  color: var(--color-notice-info--dark);
}
