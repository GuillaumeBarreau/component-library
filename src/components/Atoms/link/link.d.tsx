import { ReactNode, Ref } from "react";

export interface ComponentLinkProps {
  children: string & ReactNode;
  className?: string;
  notice?: "default" | "success" | "error" | "info" | "warning" | "disabled";
  target?: "_blank" | "_parent" | "_top" | "_seft";
  decorated?: boolean;
  href?: string;
  forwardedRef?: Ref<HTMLAnchorElement>;
}
