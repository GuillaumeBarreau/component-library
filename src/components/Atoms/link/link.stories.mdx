import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";
import { Link } from './link';
import { ComponentLinkProps } from "./link.d";

<Meta 
  title="Components/Atoms/Link" 
  component={ Link }
  argTypes={ {
    className: {
      description: "Override a component's styles",
      control: "text",
    },
    children: {
      control: "text",
      description: "Children of the link.",
    },
    notice: {
      control: "text",
      description: "Define if it should change color.",
    },
    href: {
      control: "text",
      description: "Provide the href attribute for the <a> node"
    },
  } }
  args={ {
    children: "Link children",
    className: "Override className",
    href: "/"
  } }/>

# Link

Use a Link when you want the user to navigate to another page, a specific location on a page, or to download a document.

## React Component

#### Import library

```javascript
Import {Atoms} from '@bonjour-les-pixels/component-library';
```

#### Use Link component

```javascript
<Atoms.Link {...props}> CHILDREN_TEXT </Atoms.Link>
```

## Playground

<Canvas>
  <Story name="Link">
    {args => {
      return <Link {...args}  />;
    }}
  </Story>
</Canvas>

### Props

<ArgsTable story="Link" />

### Notice

You can use the `notice` prop to modify the style color. This props can be declined in 5 ways. The style varies according to the level of importance of the action or the brand color in the page context.

<Canvas>
    <Story name="Link/Notice">
        <Link
          href="/"
          children="default link"
        />
        <br/>
        <Link
          notice="info"
          href="/"
          children="info link"
        />
        <br/>
        <Link
          notice="success"
          href="/"
          children="success link"
        />
        <br/>
        <Link
          notice="warning"
          href="/"
          children="warning link"
        />
        <br/>
        <Link
          notice="error"
          href="/"
          children="error link"
        />
     </Story>
</Canvas>

You can use the value `disabled` in `notice` prop when you need to keep a user from using the link component until some specific condition has been met.

<Canvas>
  <Link
      notice="disabled"
      href="/"
      children="link disabled"
    />
</Canvas>