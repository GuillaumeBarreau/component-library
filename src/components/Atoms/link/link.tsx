import React from "react";
import classNames from "classnames";
import { ComponentLinkProps } from "./link.d";
import "./link.css";

export const Link: React.FC<ComponentLinkProps> = (props) => {
  const {
    children,
    className,
    notice = "default",
    target = "_self",
    decorated = false,
    href,
    forwardedRef,
  } = props;

  const prefixCls = "link";
  const classes = classNames(
    prefixCls,
    {
      [`${prefixCls}--notice-${notice}`]: notice,
      [`${prefixCls}--notice-${decorated}`]: decorated,
    },
    className
  );

  return (
    <a
      ref={forwardedRef}
      className={classes}
      href={notice !== "disabled" ? href : undefined}
      target={target}
      rel={target === "_blank" ? "noopener noreferrer" : ""}
    >
      {children}
    </a>
  );
};
