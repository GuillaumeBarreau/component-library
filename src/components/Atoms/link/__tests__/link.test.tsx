import React, { Ref } from "react";
import { mount, render, shallow } from "enzyme";
import { Link } from "../link";
import { ComponentLinkProps } from "../link.d";

describe("Link", () => {
  let defaultProps: JSX.IntrinsicAttributes & ComponentLinkProps;

  beforeEach(() => {
    defaultProps = {
      children: "Link",
      notice: "info",
      href: "https://bonjoursLesPixels.com",
    };
  });

  it("renders correctly", () => {
    const wrapper = shallow(<Link {...defaultProps} />);
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("mount correctly", () => {
    const wrapper = shallow(<Link {...defaultProps} />);
    expect(() => wrapper).not.toThrow();
  });

  it("Should contain defaultProps:children correctly", () => {
    const wrapper = shallow(<Link {...defaultProps} />);
    expect(() => wrapper).not.toThrow();
    expect(wrapper.html()).toContain("Link");
  });

  it("Should contain defaultProps:href correctly", () => {
    const wrapper = shallow(<Link {...defaultProps} />);
    expect(wrapper.html()).toContain('href="https://bonjoursLesPixels.com"');
  });

  it("Should contain defaultProps:notice correctly", () => {
    const wrapper = mount(<Link {...defaultProps} />);
    expect(() => wrapper).not.toThrow();
    expect(wrapper.html()).toContain("link--notice-info");
  });

  it("Should contain Props:Notice with following value disabled", () => {
    const wrapper = shallow(<Link {...defaultProps} notice="disabled" />);
    expect(wrapper.html()).toContain("disabled");
  });
});
