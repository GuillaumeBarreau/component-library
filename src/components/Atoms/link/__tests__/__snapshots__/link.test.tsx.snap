// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders correctly 1`] = `
"<a className=\\"link link--notice-info\\" href=\\"https://bonjoursLesPixels.com\\" target=\\"_self\\" rel=\\"\\">
  Link
</a>"
`;
